⚡️ Ensure faster start-up when using many nested routers

jvllmr opened the pull request:
First of all: Thanks a lot for your on this project! I never thought that Python Web Development (especially for back-ends) could be this much more fun.

Now to the main part: In my FastAPI project I have a lot of nested routers. Because the amount of nested routers I soon reached a FastAPI start-up time of 30 seconds or more. This hurt my developer experience a lot and I soon began digging in to why FastAPI takes so long to load. 

After some research I found out that each router calculates all relevant infos on initialization. A router, which includes another router, will re-calculate these values without re-using information from the router it includes. When having a flat project structure this is not a problem. 

So the solution was simple: Defer the calculation of values until they are actually needed. By doing that the start-up time was shortened significantly. 

One caveat remains: Deferring the calculation results to Exceptions not being raised at start-up and they are only raised at the first call of a route. So they might not be caught if a route is not tested at least once.

This pull request is my proposal to integrate this performance optimization into FastAPI without breaking user space and having to modify tests. I mitigated the Exception's issue by introducing an argument to the APIRouter and APIRoute which decides if values are deferred or not and never deferring values in the top-level (app-integrated) router.

I'm sure there might be a better way. So let me hear if you have suggestions for improvement!

You can find my research and testing environment for this here: https://github.com/jvllmr/fastapi-deferred-init

Jamim commented:
Thanks you, @jvllmr!
This PR and #10546 might perfectly complement each other 🚀

jvllmr committed:
defer-route-init for better preformance with many nested routers

jvllmr commented:
I refactored the pull-request according to the latest changes.

In the meantime the performance improved already (probably because of changes in Pydantic). In my benchmark the median improvement was around 8 to 10x. My benchmark is very simple and there might be more improvement in real world applications.

The tests are failing as of now. I will look into it as soon as I can.

svlandeg committed:
Merge branch 'master' into master

svlandeg commented:
Hi @jvllmr!

> The tests are failing as of now. I will look into it as soon as I can.

There was an unrelated issue with our CI that is now fixed, so you shouldn't see so many failing tests anymore.

The only test still failing here is the one measuring "coverage" - this basically ensures that all code paths in the code base are tested in our test suite. So, to fix it, it would be good to add a test with this PR that checks the correct behaviour of the new functionality introduced here 🚀 

Specifically, when I look at the detailed coverage report, it's the `init_routes` function that isn't getting properly tested: https://smokeshow.helpmanual.io/6t005p5s2v3j6u064g0q/z_35ba7d295f00afa4_routing_py.html#t1160

jvllmr committed:
Merge branch 'master' into master

jvllmr commented:
Thanks for the hint. I will see if I can come up with some good tests over the weekend.

jvllmr committed:
recover lost status code check

jvllmr committed:
add tests for deferred route initialization

remove useless checks in test

test defer_init flag on route

add route to test app

fix coverage again

jvllmr commented:
I updated the PR with tests for all the new logic I introduced, but I don't know why 99.0% coverage are still shown here.
It was green at first but then went back to red. When looking at the detailed report it still says 100%.
I'm also sorry for the messed up commit history (squash to the rescue).

Anyway, I  added more FastAPI features to my benchmark and the improvement became even more significant: 
https://jvllmr.github.io/fastapi-deferred-init/dev/bench/

jvllmr commented:
I actually found the offending lines. For some reason the summary at the bottom showed 100%, but I found the lines immediately when checking "Hide covered". All green now ✅ 🏁

jvllmr commented:
I dropped the unintentional commits just to be sure

jvllmr committed:
Merge branch 'master' into master

svlandeg commented:
Thanks! We've got a bit of a reviewing queue at the moment, but we've added this one and will get back to you once we've been able to review it in more detail 🙏

jvllmr committed:
assert that include_router does not cause initialization

jvllmr committed:
Merge branch 'master' into master

jvllmr committed:
Merge branch 'master' into master

jvllmr committed:
Merge branch 'master' into deferred-init

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
improve formatting for and re-word doc string

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

pre-commit-ci[bot] committed:
🎨 [pre-commit.ci] Auto format from pre-commit.com hooks

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

YuriiMotov committed:
Merge remote-tracking branch 'upstream/master' into deferred-init

YuriiMotov committed:
Reduce the diff, simplify `init_attributes`, fix `create_cloned_field` comment

YuriiMotov reviewed (commented):
LGTM in general.

Let's wait for Sebastian to validate the idea, then I think we should work a bit more on tests (split by test cases, simplify a bit, maybe add more tests) and probably mention this somewhere in docs.

### The idea of this PR:
When we include the router into another, it creates copies of each `APIRoute` recursively, triggering `__init__` methods each time.
But there are some "expensive" attributes that are initialized in `__init__` and there is no point to calculate them each time we include the chain of routers (it will create a copy and trigger `__init__` again).

It makes sense to create the tree of routers first and then init those attributes once per each instance in the tree.
This PR turns those attributes into cached properties, so they are only calculated on first access (by default, but it's configurable using `defer_init` parameter).


### Performance test

| parameter | master | this pr |
|---|---|---|
| App init time | 0.0118 | 0.0052 |
| First request time | 0.0027 | 0.0024 |
| Second request time | 0.0017 | 0.0016 |
| Total time | 0.0162 | 0.0091 |

<details><summary>Test code</summary>

```py
import time

from fastapi import APIRouter, FastAPI
from fastapi.testclient import TestClient
from pydantic import BaseModel


class MyModel(BaseModel):
    id: int
    name: str


def get_app():

    router9 = APIRouter(prefix="/level9")
    @router9.post("/")
    async def read_root9(data: MyModel):
        return {"message": "Hello from Level 9"}

    router8 = APIRouter(prefix="/level8")
    router8.include_router(router9)
    @router8.post("/")
    async def read_root8(data: MyModel):
        return {"message": "Hello from Level 8"}

    router7 = APIRouter(prefix="/level7")
    router7.include_router(router8)
    @router7.post("/")
    async def read_root7(data: MyModel):
        return {"message": "Hello from Level 7"}

    router6 = APIRouter(prefix="/level6")
    router6.include_router(router7)
    @router6.post("/")
    async def read_root6(data: MyModel):
        return {"message": "Hello from Level 6"}

    router5 = APIRouter(prefix="/level5")
    router5.include_router(router6)
    @router5.post("/")
    async def read_root5(data: MyModel):
        return {"message": "Hello from Level 5"}

    router4 = APIRouter(prefix="/level4")
    router4.include_router(router5)
    @router4.post("/")
    async def read_root4(data: MyModel):
        return {"message": "Hello from Level 4"}

    router3 = APIRouter(prefix="/level3")
    router3.include_router(router4)
    @router3.post("/")
    async def read_root3(data: MyModel):
        return {"message": "Hello from Level 3"}

    router2 = APIRouter(prefix="/level2")
    router2.include_router(router3)
    @router2.post("/")
    async def read_root2(data: MyModel):
        return {"message": "Hello from Level 2"}

    router1 = APIRouter(prefix="/level1")
    router1.include_router(router2)
    @router1.post("/")
    async def read_root1(data: MyModel):
        return {"message": "Hello from Level 1"}

    app = FastAPI(title="Deeply Nested Routers Example", version="1.0")

    app.include_router(prefix="/api_1",router=router1)
    app.include_router(prefix="/api_2",router=router1)

    return app

def test():
    app_init_results = []
    first_request_results = []
    second_request_results = []
    total_time_results = []

    for _i in range(10):
        start_time = time.time()
        app = get_app()
        app_init_completion_time = time.time()
        app_init_results.append(app_init_completion_time - start_time)
        client = TestClient(app)

        response = client.post(
            "/api_1/level1/level2/level3/level4/level5/level6/level7/level8/level9",
            json={"id": 1, "name": "Test"}
        )
        assert response.status_code == 200, response.text
        first_request_completion_time = time.time()
        first_request_results.append(first_request_completion_time - app_init_completion_time)

        response = client.post(
            "/api_2/level1/level2/level3/level4/level5/level6/level7/level8/level9",
            json={"id": 1, "name": "Test"}
        )
        assert response.status_code == 200, response.text
        second_request_completion_time = time.time()
        second_request_results.append(second_request_completion_time - first_request_completion_time)

        end_time = time.time()
        duration = end_time - start_time
        total_time_results.append(duration)

    print(f"App init time: {sum(app_init_results)/len(app_init_results)}")
    print(f"First request time: {sum(first_request_results)/len(first_request_results)}")
    print(f"Second request time: {sum(second_request_results)/len(second_request_results)}")
    print(f"Total time: {sum(total_time_results)/len(total_time_results)}")


if __name__ == "__main__":
    test()
```

</details>


@jvllmr, thanks!

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
align code with master

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
update testfile

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
add missing return statement

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
fix tests

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
Merge branch 'master' into deferred-init

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
Merge branch 'master' into deferred-init

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
fix optional return type

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

pre-commit-ci-lite[bot] committed:
🎨 Auto format

jvllmr committed:
Merge branch 'master' into deferred-init

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
Merge branch 'master' into deferred-init

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
ci fixes

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
Merge branch 'master' into deferred-init

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

jvllmr committed:
remove unused comment

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

Jamim commented on fastapi/routing.py:
```diff
@@ -1002,7 +1001,7 @@ def body_field(self) -> ModelField | None:
 
     @cached_property
-    def app(self) -> ASGIApp:  # type: ignore
+    def app(self) -> ASGIApp:
```
_Hello @jvllmr,_

It's not just a comment.
This removal causes [an error on CI](https://github.com/fastapi/fastapi/actions/runs/25688976393/job/75420059622?pr=10589#step:8:27), so it's worth reverting.

jvllmr committed:
Revert "remove unused comment"

This reverts commit 6500568235f5645a5048aa53943694cb4be849b3.

jvllmr committed:
Merge branch 'master' into deferred-init

jvllmr committed:
narrow type ignore comment for mypy

Signed-off-by: Jan Vollmer <jan@vllmr.dev>

tiangolo commented:
Thanks for the interest and work on this! ☕

I want to implement a few big refactors and features that might interact with this logic.

Given I have to do these refactors soon, the risk of issues from edge cases touched by this is currently not worth it.

I have not wanted to close PRs that target problems related to things I have planned before I have an alternative, but then I have left many PRs open for too long, which is not useful or clear either. Sorry for that.

So, I'll close this one for now to keep things clean.

Thank you! 🍰
